tsdb/engine.go

```diff
@@ -489,6 +489,50 @@ func (e *Executor) processFill(results [][]interface{}) [][]interface{} {
 		return newResults
 	}
 
+	// mean fill
+	if e.stmt.Fill == influxql.MeanFill {
+		var (
+			nullFirst bool
+			nullCount = make(map[int]int)
+		)
+
+		for i, vals := range results {
+			// start at 1 because the first value is always time
+			for j := 1; j < len(vals); j++ {
+				if vals[j] == nil {
+					// record the number of nulls since the last measurement for this value
+					nullCount[j]++
+					if i == 0 {
+						nullFirst = true
+					}
+					continue
+				}
+				if nullCount[j] > 0 {
+					// back-fill null values with the mean over those windows
+					if i-nullCount[j] <= 0 {
+						// can't fill without previous value
+						nullCount[j] = 0
+						continue
+					}
+					step := (int64toFloat64(vals[j]) - int64toFloat64(results[i-nullCount[j]-1][j])) / float64(nullCount[j]+1)
+					current := int64toFloat64(vals[j])
+					var count int
+					if nullFirst {
+						count = i
```

Comment:
So, I'm not really certain if extrapolating the data back to the start like this is a good idea, or whether we should just leave the start of the series with nulls.